### DISCUSSION TITLE
Rollback and safe mode ability

### DISCUSSION THREAD

**Discussion by 737simpilot** [CIDRAM/CIDRAM#296] [General]
What if CIDRAM had some kind of rollback ability? Maybe even via date? And to complicate matters even more (LOL) :smile: , a type of "safe mode" where all none core CIDRAM modules, auxiliary rules, etc are turned off for testing purposes? 

I think it sounds prudent, but probably a major pain to code all that. Maybe something interesting for v3 if v3 has other code capabilities that v2 doesn't to allow such features. 

![img](https://i.postimg.cc/Njg6HDhn/ezgif-com-gif-maker-16.gif)

**Comment by Maikuolan** [comment]
A rollback ability would definitely be a great idea. TBH, I had thought about doing this in the past a couple of times, but the main reason why I didn't implement it in the end was because, like you mentioned.. a major pain to code. '^.^

I do already plan to completely rework the update system for v3 anyway though, so maybe this'll become something less of a pain to implement at that point (although I haven't yet started on that, basically due to time constraints all lots of other things to get around to implementing first), and I would still like to see something like that implemented, so.. we'll keep this idea in the open, I think. :-)

Regarding a "safe mode" feature: That should be possible to achieve by way of the new `fields` directive in v3. Actually.. there's quite a lot that'll be possible with that. I'd also decided to remove the `maintenance_mode` configuration directive for v3 (that much has already been committed), since it's basically obsolete now that the `fields` and `stages` directives have been implemented. :-)

But yeah.. I suspect somehow that all the changes I want to implement to the updates system is going to take a while. Should definitely be worth it in the end though.

**Comment by 737simpilot** [comment]
>but the main reason why I didn't implement it in the end was because, like you mentioned.. a major pain to code. '^.^

HAHAHA Totally understand that. Well, I've tried to learn code (C++ for what ever that's worth. I should go with C I guess, not too sure) and I immediately understood that it takes a certain logic to it all, a lot of math, and it has a very strict structure I guess you can say. On the other hand, I _might_ be able to learn it if I just dedicated myself to it because of that strict structure. I like structured things so I'm not guessing. It's like with markup or BBcode I guess. Not exactly "code" but one needs to to know the proper way to format text and what not using that code. 

I can't even imagine how many lines of code it would take. And  full well understanding the need for any debugging that would need to be done.

>Regarding a "safe mode" feature: That should be possible to achieve by way of the new fields directive in v3. Actually.. there's quite a lot that'll be possible with that.

I had a funny felling that would be the case. 

The reason why I'd like to see these two features is that I'm always testing the crap out of everything to find that one possible bug. And sometimes this involves backing up my current install and using a fresh install for the purposes of a control and a variable in my experiments. I also think it may be a nice perk to those trying to diagnose problems. Well, to those wiling to utilize it of course. :smile: 

It's probably good I don't know what the capabilities are in certain versions of PHP because I would be like, "we can do this, and that, and this!" LOL! On the other hand, if I knew PHP I'd do pull requests and contribute to the code which I wish I could do. I really do. It would take me a while though. You probably have a good five plus years experience already. And I do know it requires a lot of dedication. Like being a U.S. Navy SEAL you never give up.

**Comment by 737simpilot** [comment]
>maintenance_mode

Yeah, I remember back then (could have been while still at ZBblock's website) I requested that feature.
